normandy/recipes/exports.py

```diff
@@ -0,0 +1,207 @@
+                f"Remote Settings collection {self.collection_id} is not writable."
+            )
+
+        # Test that collection has the proper review settings.
+        capabilities = server_info["capabilities"]
+        if "signer" in capabilities:
+            signer_config = capabilities["signer"]
+            normandy_resource = [
+                r
+                for r in signer_config["resources"]
+                if r["source"]["bucket"] == settings.REMOTE_SETTINGS_BUCKET_ID
+                and r["source"]["collection"] == self.collection_id
+            ]
+            review_disabled = (
+                len(normandy_resource) == 1
+                and not normandy_resource[0].get(
+                    "to_review_enabled", signer_config["to_review_enabled"]
+                )
+                and not normandy_resource[0].get(
+                    "group_check_enabled", signer_config["group_check_enabled"]
+                )
+            )
+            if not review_disabled:
+                raise ImproperlyConfigured(
+                    f"Review was not disabled on Remote Settings collection {self.collection_id}."
+                )
+
+    def published_recipes(self):
+        """
+        Return the current list of remote records.
+        """
+        if self.client is None:
+            raise ImproperlyConfigured("Remote Settings is not enabled.")
+
+        return self.client.get_records(bucket=self.MAIN_BUCKET_ID)
+
+    def publish(self, recipe):
+        """
+        Publish the specified `recipe` on the remote server by upserting a record.
+        """
+        if self.client is None:
+            return  # no-op if disabled.
+
+        # 1. Put the record.
+        record = recipe_as_record(recipe)
+        self.client.update_record(data=record)
+        try:
+            # 2. Approve the changes immediately (multi-signoff is disabled).
+            self.client.patch_collection(data=APPROVE_CHANGES_FLAG)
+
+        except kinto_http.exceptions.KintoException:
+            # Approval failed. The changes in the `main-workspace` bucket must be reverted.
+            try:
+                # If it was an update of existing record, revert the modifications in the
+                # `main-workspace` bucket by restoring the version that was published
+                # in the `main` bucket.
+                in_prod = self.client.get_record(id=record["id"], bucket=self.MAIN_BUCKET_ID)
+                self.client.update_record(data=in_prod)
+
+            except kinto_http.exceptions.KintoException as e:
+                if e.response.status_code == 404:
+                    # It was a new record that was never published in the `main` bucket,
+                    # revert its creation in the `main-workspace` bucket by deleting it.
+                    self.client.delete_record(id=record["id"])
+            raise
+
+        logger.info(f"Published record '{recipe.id}' for recipe {recipe}")
```

Comment:
Shouldn't that be `{recipe.name}` like it is on line 207?

I for one think we should rely on repr for both these things. Feels more solid. Like:

```python
logger.info(f"Published record {recipe.id!r} for recipe {recipe.name!r}")
```

since otherwise the "name" becomes a part of the English text.